### PR TITLE
Avoid creating a tuple for varargs in some cases

### PR DESCRIPTION
**Pull request by vasslitvinov**
WE USED TO

combine varargs into a tuple the first thing in a function.
E.g when this function:

  proc test(args...) {
    BODY;
  }

is invoked with two formals, the compiler instantiates it into:

  proc test(_e0_args, _e1_args) {
    var args = _build_tuple(2, _e0_args, _e1_args);
    BODY;
  }

Indeed the vararg tuple is created in the callee, not the caller.

WITH THIS COMMIT,

the compiler no longer inserts the _build_tuple() call or creates
the local 'args' tuple when:

(A) BODY does not need 'args' as a whole tuple (see below), and
(B) the "args..." formal's intent is not 'out' or 'inout'.

This results in simpler generated code. Also, as the primary motivation
for this change, if the args formal has a param or type intent, then
(...args) consists of params or types, resp. This latter was not the case
prior to this change.

(A) details: the compiler considers (A) to hold when the only uses
of 'args' in the body, if any, are these:

  (A1) (...args)           --> replaced with _e0_args, _e1_args
  (A2) args(i) for param i --> replaced with _e0_args or _e1_args
  (A3) args.size           --> replaced with 2

The replacements shown are for the test() example above.

(B) details: supporting out/inout intent is left as future work.
These intents require formal temps. So instead of the 'args' tuple
we'd have to create individual temps. Maybe this wouldn't be
much savings to begin with?

IMPLEMENTATION NOTES

When the function whose varargs are being instantiated
had FLAG_PARTIAL_COPY, the replacements needed to be done
separately. The necessary information is passed through
the fields I added on FnSymbol.

While there, I added a missing SET_LINENO and in a different
spot set a node's astloc manually, which is more lightweight
than SET_LINENO.

ALSO ADDED TESTS

that exercise varargs in various ways.

### CODE REVIEW

File: compiler/include/symbol.h [hunk 1]
```diff
@@ -583,6 +589,9 @@ const char* intentDescrString(IntentTag intent);
 // pass-by-reference intents are used.
 bool argMustUseCPtr(Type* t);
 
+void substituteVarargTupleRefs(BlockStmt* ast, int numArgs, ArgSymbol* formal,
```
Comment (noakesmichael) [inline compiler/include/symbol.h]: This function is defined in resolution/functionResolution.cpp and so the prototype should be in
include/resolution.h.

When you move the signature please update the formatting to better handle the wrap.  As a general
rule, which is readily applied in this case, formals should either (1) fit on a single line or (2) be on
individual lines.

File: compiler/include/symbol.h [hunk 2]
```diff
@@ -353,10 +353,16 @@ class FnSymbol : public Symbol {
   int codegenUniqueNum;
   const char *doc;
 
+  // The following fields are used only when hasFlag(FLAG_PARTIAL_COPY)
```
Comment (noakesmichael) [inline compiler/include/symbol.h]: Can you please expand on the intro to the set of member variables.  It doesn't need to be long discussion
but perhaps a few sentences for the (1) the reader who isn't particularly familiar with the implications of
FLAG_PARTIAL_COPY (2) and some sort of remark to the effect that these fields suggest that the class
should be refactored.

File: compiler/AST/symbol.cpp [hunk 3]
```diff
@@ -1694,6 +1695,7 @@ void FnSymbol::finalizeCopy(void) {
       /*
        * Iterate over the statements that have been added to the function body
        * and add them to the new body.
+       * (Added only if needTupleInBody - see handleSymExprInExpandVarArgs().)
```
Comment (noakesmichael) [inline compiler/AST/symbol.cpp]: This extension to the comment doesn't really add anything unless one has a lot of
understanding about var-arg expansion.  Please either remove it or expand it.

File: compiler/AST/symbol.cpp [hunk 4]
```diff
@@ -1768,6 +1770,15 @@ void FnSymbol::finalizeCopy(void) {
      */
     update_symbols(this, map);
 
+    // Replace vararg formal if appropriate.
+    if (this->varargOldFormal) {
+      substituteVarargTupleRefs(this->body, this->varargNewFormals.size(),
```
Comment (noakesmichael) [inline compiler/AST/symbol.cpp]: The general rule of thumb, which is easily applied in this example, is to place all of the actuals on a single
line or places them on individual lines.

File: compiler/resolution/functionResolution.cpp [hunk 5]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
 }
 
 
+//
+// Is 'parent' an indexing expression into the tuple 'se'?
+// If so, return the index (between 1 and numArgs). Otherwise return 0.
+//
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: 1) There's inconsistent formatting here with respect to use of spaces before "{" and "(".

2) There an inconsistency about where the "{" appears (same line in most cases, next line in one case)

3) As a good rule of thumb, which is easily applied here, the components of the conditional for the third if
statement should either be on one line or on individual lines.

4) I think this name of this function is confusing. The pattern is<FOO> would suggest it's a predicate but then
it returns an integer. It's clear there are pros/cons to overloading the return value to mean both pass/fail and
index if pass.  If pursuing the overloading path then bias things for the expected case i.e. pass.  This is especially
easy to defend for a static function in which the caller and callee are under tight control.

File: compiler/resolution/functionResolution.cpp [hunk 6]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+// Q: Why not just keep move(dest,replFormal), which is simpler?
+// A: because 'move' does not preserv replFormal's ref-ness and const-ness,
+// so we will get errors when using dest on the l.h.s. and miss errors
+// when dest or deref(dest) is modified and it shouldn't.
+// Note that we are here before the code is resolved, so we do not
+// necessarily know whether 'replFormal' is a const and/or a ref.
+// To keep 'dest' around, instead of move we probably need to add
+// the AST equivalent of Chapel's "ref dest = replFormal;".
+//
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Formals should be on same line or individual lines

File: compiler/resolution/functionResolution.cpp [hunk 7]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+          INT_ASSERT(parent->get(2) == se);
+          return true;
+        }
+  return false;
+}
+
+//
+// Does 'ast' contain use(s) of the vararg tuple 'formal'
+// that require 'formal' as a whole tuple?
+//
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: The line "Keep in sync with substituteVarargTupleRefs", and the corresponding comment in the other function,
seem too terse to provide much value.  You seem to be implying some sort of tight-coupling between these two
functions that might not be obvious to someone who is studying the code.  This suggests the need for one slightly
longer comment that discusses this coupling that you are concerned about.

File: compiler/resolution/functionResolution.cpp [hunk 8]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+// Keep in sync with substituteVarargTupleRefs().
+//
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Formals should be on same line or individual lines

File: compiler/resolution/functionResolution.cpp [hunk 9]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+static bool needVarArgTupleAsWhole(BlockStmt* ast, int numArgs,
+                                   ArgSymbol* formal)
+{
+  std::vector<SymExpr*> symExprs;
+  collectSymExprs(ast, symExprs);
+
+  for_vector(SymExpr, se, symExprs) {
+    if (se->var == formal) {
+      if (CallExpr* parent = toCallExpr(se->parentExpr)) {
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: For my education :- what is the purpose of the SET_LINENO here?

File: compiler/resolution/functionResolution.cpp [hunk 10]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+        SET_LINENO(parent);
+        if (parent->isPrimitive(PRIM_TUPLE_EXPAND)) {
+          // (...formal) -- does not require
+          continue;
+        }
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: The overloaded nature of the return value from this function suggests value in being explicit about what is
being tested e.g if (varargAccessIndex(se, parent, numArgs) != 0) {

File: compiler/resolution/functionResolution.cpp [hunk 11]
```diff
@@ -1664,27 +1825,55 @@ handleSymExprInExpandVarArgs(FnSymbol*  workingFn,
...
-
         formal->defPoint->insertBefore(newArgDef);
+        varargFormals[i] = newFormal;
       }
 
+      bool needTupleInBody;
+
       // Replace mappings to the old formal with mappings to the new variable.
       if (workingFn->hasFlag(FLAG_PARTIAL_COPY)) {
+        bool gotFormal = false; // for assertion only
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: I recognize that this is a legal for statement but I think it is a poor decision to fold the loop update expression in to the loop test.  It's uncommon and easy to overlook.

File: compiler/resolution/functionResolution.cpp [hunk 12]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+      INT_ASSERT(false);
+    } else if (Symbol* replmt = tempReps.get(se->var)) {
+      SET_LINENO(se);
+      se->replace(new SymExpr(replmt));
+    }
+  }
+}
+
+
 /** Common code for multiple paths through expandVarArgs.
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: I recognize that this function was already unreasonably long but your edits have added to the issue.
I think we should step back for a moment and see if there's some refactoring to be done.  This seems
likely as I think that var-arg expansion should be either 1) actually use a tuple 2) avoid the tuple.  I haven't
given it a great deal of thought but I will be surprised if there isn't a way to reflect this reasonably directly in
the code.

File: compiler/resolution/functionResolution.cpp [hunk 13]
```diff
@@ -1770,8 +1962,13 @@ handleSymExprInExpandVarArgs(FnSymbol*  workingFn,
         // Otherwise, do the substitution now.
         subSymbol(workingFn->body, formal, var);
       }
+     } else {
+      // !needTupleInBody
+      substituteVarargTupleRefs(workingFn->body, n, formal, varargFormals);
+     }
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: You've added a layer of conditional but failed to update the indentation on the existing body to match.

File: compiler/include/symbol.h (hunk 1 above)
Comment (vasslitvinov) [inline compiler/include/symbol.h]: @noakesmichael - symbol.cpp does not #include resolution.h. So following your suggestion prevents building the compiler.

How do you suggest resolving this issue and how important is it to you vs. leaving substituteVarargTupleRefs() in symbol.h ?

File: compiler/resolution/functionResolution.cpp (hunk 9 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: SET_LINENO sets the file name/line number to be assign to all new AST nodes created (e.g. new SymExpr(...)) through the end of the lexical scope where SET_LINENO occurs and in all dynamically-scoped called made during this time. These file name/line number can be overridden by another SET_LINENO for the duration of its lexical scope and its dynamically-scoped calls.

File: compiler/resolution/functionResolution.cpp (hunk 11 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Given that this is pre-existing code, do you insist on fixing the issue that you point out?

File: compiler/resolution/functionResolution.cpp (hunk 13 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: I avoid messing revision history on existing code whenever I reasonably can. Given that the only issue here is indenting by 1 space instead of 2 spaces, do you insist on doing so?

File: compiler/resolution/functionResolution.cpp (hunk 12 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: I suggest doing such refactoring as a separate PR, as it is orthogonal to the change I am making here. OK with you?

File: compiler/resolution/functionResolution.cpp (hunk 9 above)
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Sorry.  I know what is does in general.  What purpose is it serving in this
function?  I did spot any allocations in that function.


On May 11, 2016, at 1:28 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62920615>:

> +
> +//
> +// Does 'ast' contain use(s) of the vararg tuple 'formal'
> +// that require 'formal' as a whole tuple?
> +//
> +// Keep in sync with substituteVarargTupleRefs().
> +//
> +static bool needVarArgTupleAsWhole(BlockStmt* ast, int numArgs,
> +                                   ArgSymbol* formal)
> +{
> +  std::vector<SymExpr*> symExprs;
> +  collectSymExprs(ast, symExprs);
> +
> +  for_vector(SymExpr, se, symExprs) {
> +    if (se->var == formal) {
> +      if (CallExpr* parent = toCallExpr(se->parentExpr)) {


SET_LINENO sets the file name/line number to be assign to all new AST nodes created (e.g. new SymExpr(...)) through the end of the lexical scope where SET_LINENO occurs and in all dynamically-scoped called made during this time. These file name/line number can be overridden by another SET_LINENO for the duration of its lexical scope and its dynamically-scoped calls.

File: compiler/resolution/functionResolution.cpp (hunk 11 above)
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Yes

On May 11, 2016, at 1:32 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62921196>:

>        // Replace mappings to the old formal with mappings to the new variable.
>        if (workingFn->hasFlag(FLAG_PARTIAL_COPY)) {
> +        bool gotFormal = false; // for assertion only


Given that this is pre-existing code, do you insist on fixing the issue that you point out?

File: compiler/resolution/functionResolution.cpp (hunk 13 above)
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: On May 11, 2016, at 1:35 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62921818>:

> @@ -1770,8 +1962,13 @@ handleSymExprInExpandVarArgs(FnSymbol*  workingFn,
>          // Otherwise, do the substitution now.
>          subSymbol(workingFn->body, formal, var);
>        }
> +     } else {
> +      // !needTupleInBody
> +      substituteVarargTupleRefs(workingFn->body, n, formal, varargFormals);
> +     }
>


I avoid messing revision history on existing code whenever I reasonably can.

1) I see the benefit but it’s more important to have consistently formatted code.
It would be silly to manage changes to the code base to favor preserving history.


2) This isn’t going to make the history any harder to inspect.



Given that the only issue here is indenting by 1 space instead of 2 spaces, do you insist on doing so?


Yes

File: compiler/resolution/functionResolution.cpp (hunk 12 above)
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: On May 11, 2016, at 1:58 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62925637>:

> +          parent->replace(new SymExpr(new_IntSymbol(numArgs)));
> +          continue;
> +        }
> +
> +      }
> +      // Cases not "continue"-ed above should have been ruled out
> +      // by needVarArgTupleAsWhole().
> +      INT_ASSERT(false);
> +    } else if (Symbol* replmt = tempReps.get(se->var)) {
> +      SET_LINENO(se);
> +      se->replace(new SymExpr(replmt));
> +    }
> +  }
> +}
> +
> +
>  /** Common code for multiple paths through expandVarArgs.


I suggest doing such refactoring as a separate PR, as it is orthogonal to the change I am making here. OK with you?


I don’t think that it’s orthogonal.  I think this work has (a) exposed an existing challenge
in the code base (b) exacerbated the problem.  That’s pretty common of course.


I agree that it’s great to do factorings in small pieces when practical and I
would have attempted that too.   To be clear the path would be

1) Start to do some development of a new component.

2) Realize the integration is going to be challenging.

3) Do the appropriate non-functional refactorings to enable the
original goal.

4) Apply the new component to the refactored code.


The alternative order tends to lead to an increasing tangle of code with
comments that say “come back and clean this up”.  A bit like our conversation
on generic function instatiation.


What are you proposing?

Mike

File: compiler/resolution/functionResolution.cpp (hunk 9 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: No purpose evidently. I deleted it.

File: compiler/resolution/functionResolution.cpp (hunk 12 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Given that you added 3x as much code to this function in #3231 without bothering to refactor it, I suggest to make it a shared endeavor and, yes, do it as a separate PR. Which I would take on right after merging this PR.

Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: On May 11, 2016, at 3:30 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62938475>:

> +          parent->replace(new SymExpr(new_IntSymbol(numArgs)));
> +          continue;
> +        }
> +
> +      }
> +      // Cases not "continue"-ed above should have been ruled out
> +      // by needVarArgTupleAsWhole().
> +      INT_ASSERT(false);
> +    } else if (Symbol* replmt = tempReps.get(se->var)) {
> +      SET_LINENO(se);
> +      se->replace(new SymExpr(replmt));
> +    }
> +  }
> +}
> +
> +
>  /** Common code for multiple paths through expandVarArgs.


Given that you added 3x as much code to this function in #3231<https://github.com/chapel-lang/chapel/pull/3231> without bothering to refactor it,

It’s not that I didn’t “bother” to refactor it.  It’s that I failed to see the need.  My error.  Of course
2 wrongs don’t make a right.




I suggest to make it a shared endeavor and, yes, do it as a separate PR. Which I would take on right after merging this PR.


I can live with that.

Mike

File: compiler/AST/symbol.cpp [hunk 14]
```diff
@@ -1768,6 +1770,15 @@ void FnSymbol::finalizeCopy(void) {
...
+                                this->varargOldFormal, this->varargNewFormals);
+      // Done, clean up.
```
Comment (noakesmichael) [inline compiler/AST/symbol.cpp]: this seems to be unchanged.

File: compiler/include/symbol.h [hunk 15]
```diff
@@ -353,10 +353,18 @@ class FnSymbol : public Symbol {
...
+  // to pass information from partialCopy() to finalizeCopy().
+  // This localized use suggests that they should be moved to a hashtable.
```
Comment (noakesmichael) [inline compiler/include/symbol.h]: Please remove the last line.  I don't think that's the right path and
wouldn't want to lead someone astray.

File: compiler/resolution/functionResolution.cpp [hunk 16]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+// Q: Why not just keep move(dest,replFormal), which is simpler?
+// A: because 'move' does not preserv replFormal's ref-ness and const-ness,
+// so we will get errors when using dest on the l.h.s. and miss errors
+// when dest or deref(dest) is modified and it shouldn't.
+// Note that we are here before the code is resolved, so we do not
+// necessarily know whether 'replFormal' is a const and/or a ref.
+// To keep 'dest' around, instead of move we probably need to add
+// the AST equivalent of Chapel's "ref dest = replFormal;".
+//
+static void replaceVarargAccess(CallExpr* parent, SymExpr* se,
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: The formal should be on 1 line or 4 (presumably 4).

File: compiler/resolution/functionResolution.cpp [hunk 17]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+  return false;
+}
+
+//
+// Does 'ast' contain use(s) of the vararg tuple 'formal'
+// that require 'formal' as a whole tuple?
+//
+// needVarArgTupleAsWhole() and substituteVarargTupleRefs() should handle
+// the exact same set of SymExprs.
+//
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Formals should be on 1 line or N.  Can't tell if this would fit on 1 line or not in this format.

File: compiler/resolution/functionResolution.cpp [hunk 18]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+  return false;
+}
+
+//
+// Replace, in 'ast', uses of the vararg tuple 'formal'
+// with references to individual formals.
+//
+// needVarArgTupleAsWhole() and substituteVarargTupleRefs() should handle
+// the exact same set of SymExprs.
+//
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Formals should ....

File: compiler/resolution/functionResolution.cpp [hunk 19]
```diff
@@ -1706,6 +1895,9 @@ handleSymExprInExpandVarArgs(FnSymbol*  workingFn,
         }
       }
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: It's hard to tell in a browser but I think the indentation is off here.  My guess is this is another
instance in which you've inserted a conditional but hesitated to re-indent correctly.

File: compiler/include/resolution.h [hunk 20]
```diff
@@ -47,6 +48,7 @@ const char* toString(CallInfo* info);
 const char* toString(FnSymbol* fn);
 
 void parseExplainFlag(char* flag, int* line, ModuleSymbol** module);
```
Comment (noakesmichael) [inline compiler/include/resolution.h]: The signature exceeds the 80 column guideline.  Should be on 4 lines.

File: compiler/resolution/functionResolution.cpp [hunk 21]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+}
+
+//
+// Does 'ast' contain use(s) of the vararg tuple 'formal'
+// that require 'formal' as a whole tuple?
+//
+// needVarArgTupleAsWhole() and substituteVarargTupleRefs() should handle
+// the exact same set of SymExprs.
+//
+static bool needVarArgTupleAsWhole(BlockStmt* ast, int numArgs,
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Inconsistent placement for the opening "{"

File: compiler/resolution/functionResolution.cpp [hunk 22]
```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
...
+}
+
+//
+// Replace, in 'ast', uses of the vararg tuple 'formal'
+// with references to individual formals.
+//
+// needVarArgTupleAsWhole() and substituteVarargTupleRefs() should handle
+// the exact same set of SymExprs.
+//
+void substituteVarargTupleRefs(BlockStmt* ast, int numArgs, ArgSymbol* formal,
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Inconsistent placement for first "{"

File: compiler/resolution/functionResolution.cpp (hunk 19 above)
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp]: @vasslitvinov - note that a reviewer can hide whitespaces changes when reviewing if they append `?w=l` to the url.  I personally wouldn't avoid making whitespace changes when adding code like this - it's more important to be able to read the file on its own cleanly than avoid an extra step when tracking down changes to that line.

File: compiler/AST/symbol.cpp (hunk 4 above)
Comment (vasslitvinov) [inline compiler/AST/symbol.cpp]: I do not know this rule. We can assume that diverse formatting styles are acceptable.

File: compiler/include/resolution.h (hunk 20 above)
Comment (vasslitvinov) [inline compiler/include/resolution.h]: This is consistent with the rest of the file.

File: compiler/resolution/functionResolution.cpp (hunk 21 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Adjusted on isVarargSizeExpr() for consistency.

File: compiler/AST/symbol.cpp (hunk 14 above)
Comment (vasslitvinov) [inline compiler/AST/symbol.cpp]: Correct.

File: compiler/include/symbol.h (hunk 15 above)
Comment (vasslitvinov) [inline compiler/include/symbol.h]: Done.

File: compiler/resolution/functionResolution.cpp (hunk 16 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: I am using a different formatting style.

File: compiler/resolution/functionResolution.cpp (hunk 17 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Ditto.

File: compiler/resolution/functionResolution.cpp (hunk 18 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: ... ditto.

File: compiler/resolution/functionResolution.cpp (hunk 22 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: It is consistent now.

File: compiler/resolution/functionResolution.cpp (hunk 19 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: I adjusted the indentation.

File: compiler/resolution/functionResolution.cpp (hunk 7 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Done.

File: compiler/resolution/functionResolution.cpp (hunk 6 above)
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: Formals should be on one line or N individual lines

File: compiler/resolution/functionResolution.cpp (hunk 17 above)
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: N Formals should be on one line or N lines

File: compiler/resolution/functionResolution.cpp [hunk 23]
```diff
@@ -1664,27 +1826,55 @@ handleSymExprInExpandVarArgs(FnSymbol*  workingFn,
...
       if (workingFn->hasFlag(FLAG_PARTIAL_COPY)) {
-        for (int index = workingFn->partialCopyMap.n; --index >= 0;) {
+        bool gotFormal = false; // for assertion only
+        for (int index = workingFn->partialCopyMap.n-1; index >= 0; index--) {
           SymbolMapElem& mapElem = workingFn->partialCopyMap.v[index];
 
           if (mapElem.value == formal) {
-            mapElem.value = var;
+            gotFormal = true;
+            needTupleInBody =
```
Comment (noakesmichael) [inline compiler/resolution/functionResolution.cpp]: N actuals to a call should be on one line or on N lines

File: compiler/AST/symbol.cpp [hunk 24]
```diff
@@ -1768,6 +1770,15 @@ void FnSymbol::finalizeCopy(void) {
      */
     update_symbols(this, map);
 
+    // Replace vararg formal if appropriate.
+    if (this->varargOldFormal) {
```
Comment (noakesmichael) [inline compiler/AST/symbol.cpp]: N actuals should be on 1 line or N lines

File: compiler/include/resolution.h (hunk 20 above)
Comment (noakesmichael) [inline compiler/include/resolution.h]: The signature exceeds the 80 column guideline.  Should be on 4 lines.

**Comment by vasslitvinov** [comment]
Superseded by #3930 .
